glean-core/src/metrics/labeled.rs

```diff
@@ -67,8 +69,30 @@ where
         }
     }
 
-    fn dynamic_label<'a>(&mut self, label: &'a str) -> &'a str {
-        // TODO(bug 1554970): Fetch seen_labels from the database if empty
+    fn dynamic_label<'a>(&mut self, glean: &Glean, label: &'a str) -> &'a str {
+        if self.seen_labels.is_empty() && self.meta.lifetime != Lifetime::Application {
+            // Fetch all labels that are already stored by iterating through existing data.
+
+            let prefix = format!("{}/", self.meta.identifier());
+            let mut seen_labels = HashSet::new();
+            let mut snapshotter = |metric_name: &[u8], _: &Metric| {
+                let metric_name = String::from_utf8_lossy(metric_name);
+                if metric_name.starts_with(&prefix) {
+                    let label = metric_name.splitn(2, '/').nth(1).unwrap(); // safe unwrap, we know it contains a slash
+                    seen_labels.insert(label.into());
+                }
+            };
+
+            let lifetime = self.meta.lifetime;
+            for store in &self.meta.send_in_pings {
+                glean
+                    .storage()
+                    .iter_store_from(lifetime, store, &mut snapshotter);
+            }
+
+            // Swap the newly seen labels with the old (empty) storage.
+            mem::swap(&mut self.seen_labels, &mut seen_labels);
```

Comment:
Learning question: Is this just so this is transactional, or is there a ownership reason to do this?

Reply:
Actually ... it was for ownership reasons, but I now see how to refactor it.